synapse/handlers/profile.py

```diff
@@ -38,11 +43,78 @@ def __init__(self, hs):
 
         self.user_directory_handler = hs.get_user_directory_handler()
 
+        self.http_client = hs.get_simple_http_client()
+
+        self.clock.looping_call(self._update_remote_profile_cache, self.PROFILE_UPDATE_MS)
+
         if hs.config.worker_app is None:
             self.clock.looping_call(
                 self._update_remote_profile_cache, self.PROFILE_UPDATE_MS,
             )
 
+        reactor.callWhenRunning(self._assign_profile_replication_batches)
+        reactor.callWhenRunning(self._replicate_profiles)
+        self.clock.looping_call(self._replicate_profiles, self.PROFILE_REPLICATE_INTERVAL)
+
+    @defer.inlineCallbacks
+    def _assign_profile_replication_batches(self):
+        """If no profile replication has been done yet, allocate replication batch
+        numbers to each profile to start the replication process.
+        """
+        logger.info("Assigning profile batch numbers...")
+        total = 0
+        while True:
+            assigned = yield self.store.assign_profile_batch()
+            total += assigned
+            if assigned == 0:
+                break
+        logger.info("Assigned %d profile batch numbers", total)
+
+    @defer.inlineCallbacks
+    def _replicate_profiles(self):
+        """If any profile data has been updated and not pushed to the replication targets,
+        replicate it.
+        """
+        host_batches = yield self.store.get_replication_hosts()
+        latest_batch = yield self.store.get_latest_profile_replication_batch_number()
+        for repl_host in self.hs.config.replicate_user_profiles_to:
+            if repl_host not in host_batches:
+                host_batches[repl_host] = -1
+            try:
+                for i in xrange(host_batches[repl_host] + 1, latest_batch + 1):
+                    yield self._replicate_host_profile_batch(repl_host, i)
+            except:
+                logger.exception(
+                    "Exception while replicating to %s: aborting for now", repl_host,
+                )
+
+    @defer.inlineCallbacks
+    def _replicate_host_profile_batch(self, host, batchnum):
+        logger.info("Replicating profile batch %d to %s", batchnum, host)
+        batch_rows = yield self.store.get_profile_batch(batchnum)
+        batch = {
+            UserID(r["user_id"], self.hs.hostname).to_string(): {
+                "display_name": r["displayname"],
+                "avatar_url": r["avatar_url"],
+            } for r in batch_rows
+        }
+
+        url = "https://%s/_matrix/federation/v1/replicate_profiles" % (host,)
```

Comment:
There doesn't seem to be a handler for this. Are profiles intended to be replicated to a non-homeserver?

Reply:
correct; this is an experiment in replicating user profiles over to the IS so that for private IS deployments the IS can act as a global address book.